sierra_adapter/sierra_reader_new/sierra_reader.py

```diff
@@ -0,0 +1,173 @@
+import json
+import os
+import uuid
+
+import boto3
+
+from client import catalogue_client
+
+
+def get_windows(event):
+    for record in event["Records"]:
+        yield json.loads(record["Sns"]["Message"])
+
+
+def get_sierra_records(client, window):
+    """
+    Fetch the records from Sierra for the given window.
+    """
+    resource_type = os.environ["RESOURCE_TYPE"]
+    sierra_fields = os.environ["SIERRA_FIELDS"]
+
+    # The updatedDate in Sierra is a timestamp supporting second-level
+    # granularity, so we want a query like this:
+    #
+    #       updatedDate=[2022-07-16T23:59:00+00:00,2022-07-20T00:01:01+00:00]
+    #
+    # This only fetches Sierra records which have been updated but not
+    # deleted.
+    print(f"Fetching updated records for {window}")
+    for record in client.get_objects(
+        f"/{resource_type}",
+        params={
+            "fields": sierra_fields,
+            "updatedDate": f'[{window["start"]},{window["end"]}]',
+        },
+    ):
+        yield record
+
+    dates = {window["start"].split("T")[0], window["end"].split("T")[0]}
+
+    # Because filtering by updatedDate doesn't catch deleted records, we
```

Comment:
Yeah, this is a tricky one. I think we do want to pick up deletions on the same interval as regular windows – from a Collections Information perspective, they want deletions to propagate as quickly as updates – but this model means we'll repeatedly process a deleted record, the first time we see it and then every time after.

I'm thinking of maybe tracking deletions somewhere in the Lambda context, so the Lambda doesn't repeatedly re-send them. And then log the time when something was initially spotted as deleted, so we know when in the day it occurred – rather than the end of the day.

Reply:
I agree it's important that we process them promptly, and if it's the same interval as updates, then it might as well be in the same process.

I think re-processing deletions is fine, as a deletion will be idempotent (unless attempting to delete an absent record yields an exception) and NOOP deletions should be pretty speedy.

I'd also expect deletions to be low volume. If we end up deleting the same few records several times in a day, then it's not going to break the bank or fill up logs or anything.

Reply:
> I think re-processing deletions is fine, as a deletion will be idempotent (unless attempting to delete an absent record yields an exception) and NOOP deletions should be pretty speedy.

Good point, the whole idea of this rewrite was to reduce the premature optimisations in the Sierra reader. 😅 

I'll put some comments about this in the code.